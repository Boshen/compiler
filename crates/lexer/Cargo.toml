[package]
name = "lexer"
version = "0.0.0"
edition = "2021"

[dependencies]
unicode-id = "0.3.0"

[dev-dependencies]
insta = "1.12.0"
unindent = "0.1.7"
