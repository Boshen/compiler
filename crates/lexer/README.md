# Lexer

## Goal

### Performance
Performance should be achieved by reducing the total amount of if checks in any code path.

## Fuzzing

```bash
cd crates/lexer && cargo +nightly fuzz run lexer -- -only_ascii=1
```
