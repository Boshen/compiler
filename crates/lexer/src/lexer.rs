//! Lexer

use crate::kind::Kind;
use crate::token::Token;

pub struct Lexer<'a> {
    /// The input string
    bytes: &'a [u8],

    /// The current index position
    cur: usize,

    /// Are we at the EOF?
    eof: bool,
}

impl<'a> Lexer<'a> {
    #[must_use]
    pub const fn new(input: &'a str) -> Self {
        Self {
            bytes: input.as_bytes(),
            cur: 0,
            eof: false,
        }
    }
}

impl Iterator for Lexer<'_> {
    type Item = Token;

    fn next(&mut self) -> Option<Self::Item> {
        if self.cur >= self.bytes.len() {
            if self.eof {
                return None;
            }
            self.eof = true;
            return Some(Token::new(Kind::EOF, self.cur, 1));
        }

        let token = self
            .read_whitespace()
            .or_else(|| self.read_identifier())
            .or_else(|| Some(Token::new(Kind::Unknown, self.cur, 1)));

        self.cur += token.as_ref().unwrap().len();

        token
    }
}

impl Lexer<'_> {
    fn read_whitespace(&mut self) -> Option<Token> {
        let mut len = 0;

        for chr in &self.bytes[self.cur..] {
            if !chr.is_ascii_whitespace() {
                break;
            }
            len += 1;
        }

        if len == 0 {
            return None;
        }

        Some(Token::new(Kind::WhiteSpace, self.cur, len))
    }

    fn read_identifier(&mut self) -> Option<Token> {
        let mut len = 0;

        for chr in &self.bytes[self.cur..] {
            if !chr.is_ascii_alphabetic() {
                break;
            }
            len += 1;
        }

        if len == 0 {
            return None;
        }

        let kind = match &self.bytes[self.cur..self.cur + len] {
            b"null" => Kind::Null,
            b"undefined" => Kind::Undefined,
            b"true" => Kind::True,
            b"false" => Kind::False,
            _ => Kind::Ident,
        };

        Some(Token::new(kind, self.cur, len))
    }
}
