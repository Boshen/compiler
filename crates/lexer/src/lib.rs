#![warn(clippy::all, clippy::pedantic, clippy::nursery)]

mod constants;
mod kind;
mod lexer;
mod state;
mod token;

pub use crate::kind::*;
pub use crate::lexer::*;
pub use crate::token::*;
