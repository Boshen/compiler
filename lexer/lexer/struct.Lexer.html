<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=../../lexer/struct.Lexer.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="../../lexer/struct.Lexer.html">../../lexer/struct.Lexer.html</a>...</p>
    <script>location.replace("../../lexer/struct.Lexer.html" + location.search + location.hash);</script>
</body>
</html>