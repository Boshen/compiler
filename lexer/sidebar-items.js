initSidebarItems({"enum":[["Kind",""],["Number",""]],"struct":[["Lexer",""],["Token",""]]});