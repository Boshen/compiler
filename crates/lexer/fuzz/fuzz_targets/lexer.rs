#![no_main]

use lexer::Lexer;
use libfuzzer_sys::fuzz_target;

fuzz_target!(|data: &[u8]| {
    if let Ok(s) = std::str::from_utf8(data) {
        Lexer::new(s).into_iter().count();
    }
});
