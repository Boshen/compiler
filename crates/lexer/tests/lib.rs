#![warn(clippy::all, clippy::pedantic, clippy::nursery)]
#[cfg(test)]
use lexer::Lexer;
use unindent::unindent;

fn test_snapshot(name: &str, input: &str) {
    let input = unindent(input);
    let tokens = Lexer::new(input.trim())
        .into_iter()
        .filter(|t| !t.kind().is_whitespace())
        .collect::<Vec<_>>();
    let snapshot = format!("# Input\n{}\n---\n# Output\n{:#?}", input, tokens);
    insta::with_settings!({
        prepend_module_to_snapshot => false,
    }, {
        insta::assert_snapshot!(name, snapshot, name);
    });
}

#[test]
fn comment() {
    let input = "
        // comment
        /* multiline comment */
    ";
    test_snapshot("comment", input);
}

#[test]
fn keyword() {
    let input = "
        await break case catch class const continue debugger default delete do else enum export extends
        false finally for function if import in instanceof new null return super switch this throw true try typeof var void while with yield
        undefined
    ";
    test_snapshot("keyword", input);
}

#[test]
fn identifier() {
    let input = r#"
        $ _ $a _a abc_$
        \u03bc
        \u{61}
    "#;
    test_snapshot("identifier", input);
}

#[test]
fn punctuator() {
    let input = "
        { ( ) [ ] . ... ; , < > <= >= == != === !== + - * % ** ++ -- << >> >>> & | ^ ! ~ && || ?? ? : = +=
        -= *= %= **= <<= >>= >>>= &= |= ^= &&= ||= ??= =>
        ?. / /= }
        #
    ";
    test_snapshot("punctuator", input);
}

#[test]
fn numeric_literal() {
    let input = "
        0 0n
        0b1 0B12
        0o1 0O12
        0x1 0X12
        0123 0789
        0. 0.0 0.123
        123n
        1.0 1.1
    ";
    test_snapshot("numeric_literal", input);
}

#[test]
fn string_literal() {
    let input = r#"
        "12345" '12345'
        "" ''
        '\\\n\r\t\b\v\f\'\"'
        '\\\n\r\t\b\v\f\'\"'
        '\u1234' '\x12'
        'foo \
        '
        "\d"
    "#;
    test_snapshot("string_literal", input);
}

#[test]
fn regex() {
    let input = r#"
        /aa/;
        /[0-9A-Za-z_\$(|)\[\]\/\\^]/;
        /[//]/;
        /[/]/;
    "#;
    test_snapshot("regex", input);
}

#[test]
fn template_literal() {
    let input = r#"
        ``
        `123`
        `\`\r`
    "#;
    test_snapshot("template_literal", input);
}
