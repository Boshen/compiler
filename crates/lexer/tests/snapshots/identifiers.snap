---
source: crates/lexer/tests/lib.rs
assertion_line: 12
expression: identifiers

---
# Input

    // comment
    /* multiline comment */
    await break case catch class const continue debugger default delete do else enum export extends
false finally for function if import in instanceof new null return super switch this throw true try typeof var void while with yield
    undefined
    $ _ $a _a abc_$
    { ( ) [ ] . ... ; , < > <= >= == != === !== + - * % ** ++ -- << >> >>> & | ^ ! ~ && || ?? ? : = +=
-= *= %= **= <<= >>= >>>= &= |= ^= &&= ||= ??= =>
    ?. / /= }
    
---
# Output
[
    Token {
        kind: Comment,
        offset: 0,
        len: 11,
    },
    Token {
        kind: WhiteSpace,
        offset: 11,
        len: 4,
    },
    Token {
        kind: MultilineComment,
        offset: 15,
        len: 23,
    },
    Token {
        kind: WhiteSpace,
        offset: 38,
        len: 5,
    },
    Token {
        kind: Await,
        offset: 43,
        len: 5,
    },
    Token {
        kind: WhiteSpace,
        offset: 48,
        len: 1,
    },
    Token {
        kind: Break,
        offset: 49,
        len: 5,
    },
    Token {
        kind: WhiteSpace,
        offset: 54,
        len: 1,
    },
    Token {
        kind: Case,
        offset: 55,
        len: 4,
    },
    Token {
        kind: WhiteSpace,
        offset: 59,
        len: 1,
    },
    Token {
        kind: Catch,
        offset: 60,
        len: 5,
    },
    Token {
        kind: WhiteSpace,
        offset: 65,
        len: 1,
    },
    Token {
        kind: Class,
        offset: 66,
        len: 5,
    },
    Token {
        kind: WhiteSpace,
        offset: 71,
        len: 1,
    },
    Token {
        kind: Const,
        offset: 72,
        len: 5,
    },
    Token {
        kind: WhiteSpace,
        offset: 77,
        len: 1,
    },
    Token {
        kind: Continue,
        offset: 78,
        len: 8,
    },
    Token {
        kind: WhiteSpace,
        offset: 86,
        len: 1,
    },
    Token {
        kind: Debugger,
        offset: 87,
        len: 8,
    },
    Token {
        kind: WhiteSpace,
        offset: 95,
        len: 1,
    },
    Token {
        kind: DefaulT,
        offset: 96,
        len: 7,
    },
    Token {
        kind: WhiteSpace,
        offset: 103,
        len: 1,
    },
    Token {
        kind: Delete,
        offset: 104,
        len: 6,
    },
    Token {
        kind: WhiteSpace,
        offset: 110,
        len: 1,
    },
    Token {
        kind: Do,
        offset: 111,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 113,
        len: 1,
    },
    Token {
        kind: Else,
        offset: 114,
        len: 4,
    },
    Token {
        kind: WhiteSpace,
        offset: 118,
        len: 1,
    },
    Token {
        kind: Enum,
        offset: 119,
        len: 4,
    },
    Token {
        kind: WhiteSpace,
        offset: 123,
        len: 1,
    },
    Token {
        kind: Export,
        offset: 124,
        len: 6,
    },
    Token {
        kind: WhiteSpace,
        offset: 130,
        len: 1,
    },
    Token {
        kind: Extends,
        offset: 131,
        len: 7,
    },
    Token {
        kind: WhiteSpace,
        offset: 138,
        len: 1,
    },
    Token {
        kind: False,
        offset: 139,
        len: 5,
    },
    Token {
        kind: WhiteSpace,
        offset: 144,
        len: 1,
    },
    Token {
        kind: FinallY,
        offset: 145,
        len: 7,
    },
    Token {
        kind: WhiteSpace,
        offset: 152,
        len: 1,
    },
    Token {
        kind: For,
        offset: 153,
        len: 3,
    },
    Token {
        kind: WhiteSpace,
        offset: 156,
        len: 1,
    },
    Token {
        kind: Function,
        offset: 157,
        len: 8,
    },
    Token {
        kind: WhiteSpace,
        offset: 165,
        len: 1,
    },
    Token {
        kind: If,
        offset: 166,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 168,
        len: 1,
    },
    Token {
        kind: Import,
        offset: 169,
        len: 6,
    },
    Token {
        kind: WhiteSpace,
        offset: 175,
        len: 1,
    },
    Token {
        kind: In,
        offset: 176,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 178,
        len: 1,
    },
    Token {
        kind: Instanceof,
        offset: 179,
        len: 10,
    },
    Token {
        kind: WhiteSpace,
        offset: 189,
        len: 1,
    },
    Token {
        kind: New,
        offset: 190,
        len: 3,
    },
    Token {
        kind: WhiteSpace,
        offset: 193,
        len: 1,
    },
    Token {
        kind: Null,
        offset: 194,
        len: 4,
    },
    Token {
        kind: WhiteSpace,
        offset: 198,
        len: 1,
    },
    Token {
        kind: Return,
        offset: 199,
        len: 6,
    },
    Token {
        kind: WhiteSpace,
        offset: 205,
        len: 1,
    },
    Token {
        kind: Super,
        offset: 206,
        len: 5,
    },
    Token {
        kind: WhiteSpace,
        offset: 211,
        len: 1,
    },
    Token {
        kind: Switch,
        offset: 212,
        len: 6,
    },
    Token {
        kind: WhiteSpace,
        offset: 218,
        len: 1,
    },
    Token {
        kind: This,
        offset: 219,
        len: 4,
    },
    Token {
        kind: WhiteSpace,
        offset: 223,
        len: 1,
    },
    Token {
        kind: Throw,
        offset: 224,
        len: 5,
    },
    Token {
        kind: WhiteSpace,
        offset: 229,
        len: 1,
    },
    Token {
        kind: True,
        offset: 230,
        len: 4,
    },
    Token {
        kind: WhiteSpace,
        offset: 234,
        len: 1,
    },
    Token {
        kind: Try,
        offset: 235,
        len: 3,
    },
    Token {
        kind: WhiteSpace,
        offset: 238,
        len: 1,
    },
    Token {
        kind: Typeof,
        offset: 239,
        len: 6,
    },
    Token {
        kind: WhiteSpace,
        offset: 245,
        len: 1,
    },
    Token {
        kind: Var,
        offset: 246,
        len: 3,
    },
    Token {
        kind: WhiteSpace,
        offset: 249,
        len: 1,
    },
    Token {
        kind: Void,
        offset: 250,
        len: 4,
    },
    Token {
        kind: WhiteSpace,
        offset: 254,
        len: 1,
    },
    Token {
        kind: While,
        offset: 255,
        len: 5,
    },
    Token {
        kind: WhiteSpace,
        offset: 260,
        len: 1,
    },
    Token {
        kind: With,
        offset: 261,
        len: 4,
    },
    Token {
        kind: WhiteSpace,
        offset: 265,
        len: 1,
    },
    Token {
        kind: Yield,
        offset: 266,
        len: 5,
    },
    Token {
        kind: WhiteSpace,
        offset: 271,
        len: 5,
    },
    Token {
        kind: Ident,
        offset: 276,
        len: 9,
    },
    Token {
        kind: WhiteSpace,
        offset: 285,
        len: 5,
    },
    Token {
        kind: Ident,
        offset: 290,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 291,
        len: 1,
    },
    Token {
        kind: Ident,
        offset: 292,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 293,
        len: 1,
    },
    Token {
        kind: Ident,
        offset: 294,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 296,
        len: 1,
    },
    Token {
        kind: Ident,
        offset: 297,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 299,
        len: 1,
    },
    Token {
        kind: Ident,
        offset: 300,
        len: 5,
    },
    Token {
        kind: WhiteSpace,
        offset: 305,
        len: 5,
    },
    Token {
        kind: LCurly,
        offset: 310,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 311,
        len: 1,
    },
    Token {
        kind: LParen,
        offset: 312,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 313,
        len: 1,
    },
    Token {
        kind: RParen,
        offset: 314,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 315,
        len: 1,
    },
    Token {
        kind: LBrack,
        offset: 316,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 317,
        len: 1,
    },
    Token {
        kind: RBrack,
        offset: 318,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 319,
        len: 1,
    },
    Token {
        kind: Dot,
        offset: 320,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 321,
        len: 1,
    },
    Token {
        kind: Dot3,
        offset: 322,
        len: 3,
    },
    Token {
        kind: WhiteSpace,
        offset: 325,
        len: 1,
    },
    Token {
        kind: Semicolon,
        offset: 326,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 327,
        len: 1,
    },
    Token {
        kind: Comma,
        offset: 328,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 329,
        len: 1,
    },
    Token {
        kind: LAngle,
        offset: 330,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 331,
        len: 1,
    },
    Token {
        kind: RAngle,
        offset: 332,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 333,
        len: 1,
    },
    Token {
        kind: LtEq,
        offset: 334,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 336,
        len: 1,
    },
    Token {
        kind: GtEq,
        offset: 337,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 339,
        len: 1,
    },
    Token {
        kind: Eq2,
        offset: 340,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 342,
        len: 1,
    },
    Token {
        kind: Neq,
        offset: 343,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 345,
        len: 1,
    },
    Token {
        kind: Eq3,
        offset: 346,
        len: 3,
    },
    Token {
        kind: WhiteSpace,
        offset: 349,
        len: 1,
    },
    Token {
        kind: Neq2,
        offset: 350,
        len: 3,
    },
    Token {
        kind: WhiteSpace,
        offset: 353,
        len: 1,
    },
    Token {
        kind: Plus,
        offset: 354,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 355,
        len: 1,
    },
    Token {
        kind: Minus,
        offset: 356,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 357,
        len: 1,
    },
    Token {
        kind: Star,
        offset: 358,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 359,
        len: 1,
    },
    Token {
        kind: Percent,
        offset: 360,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 361,
        len: 1,
    },
    Token {
        kind: Star2,
        offset: 362,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 364,
        len: 1,
    },
    Token {
        kind: Plus2,
        offset: 365,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 367,
        len: 1,
    },
    Token {
        kind: Minus2,
        offset: 368,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 370,
        len: 1,
    },
    Token {
        kind: ShiftLeft,
        offset: 371,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 373,
        len: 1,
    },
    Token {
        kind: ShiftRight,
        offset: 374,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 376,
        len: 1,
    },
    Token {
        kind: ShiftRight3,
        offset: 377,
        len: 3,
    },
    Token {
        kind: WhiteSpace,
        offset: 380,
        len: 1,
    },
    Token {
        kind: Amp,
        offset: 381,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 382,
        len: 1,
    },
    Token {
        kind: Pipe,
        offset: 383,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 384,
        len: 1,
    },
    Token {
        kind: Caret,
        offset: 385,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 386,
        len: 1,
    },
    Token {
        kind: Bang,
        offset: 387,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 388,
        len: 1,
    },
    Token {
        kind: Tilde,
        offset: 389,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 390,
        len: 1,
    },
    Token {
        kind: Amp2,
        offset: 391,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 393,
        len: 1,
    },
    Token {
        kind: Pipe2,
        offset: 394,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 396,
        len: 1,
    },
    Token {
        kind: Question2,
        offset: 397,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 399,
        len: 1,
    },
    Token {
        kind: Question,
        offset: 400,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 401,
        len: 1,
    },
    Token {
        kind: Colon,
        offset: 402,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 403,
        len: 1,
    },
    Token {
        kind: Eq,
        offset: 404,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 405,
        len: 1,
    },
    Token {
        kind: PlusEq,
        offset: 406,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 408,
        len: 1,
    },
    Token {
        kind: MinusEq,
        offset: 409,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 411,
        len: 1,
    },
    Token {
        kind: StarEq,
        offset: 412,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 414,
        len: 1,
    },
    Token {
        kind: PercentEq,
        offset: 415,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 417,
        len: 1,
    },
    Token {
        kind: Star2Eq,
        offset: 418,
        len: 3,
    },
    Token {
        kind: WhiteSpace,
        offset: 421,
        len: 1,
    },
    Token {
        kind: ShiftLeftEq,
        offset: 422,
        len: 3,
    },
    Token {
        kind: WhiteSpace,
        offset: 425,
        len: 1,
    },
    Token {
        kind: ShiftRightEq,
        offset: 426,
        len: 3,
    },
    Token {
        kind: WhiteSpace,
        offset: 429,
        len: 1,
    },
    Token {
        kind: ShiftRight3Eq,
        offset: 430,
        len: 4,
    },
    Token {
        kind: WhiteSpace,
        offset: 434,
        len: 1,
    },
    Token {
        kind: AmpEq,
        offset: 435,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 437,
        len: 1,
    },
    Token {
        kind: PipeEq,
        offset: 438,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 440,
        len: 1,
    },
    Token {
        kind: CaretEq,
        offset: 441,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 443,
        len: 1,
    },
    Token {
        kind: Amp2Eq,
        offset: 444,
        len: 3,
    },
    Token {
        kind: WhiteSpace,
        offset: 447,
        len: 1,
    },
    Token {
        kind: Pipe2Eq,
        offset: 448,
        len: 3,
    },
    Token {
        kind: WhiteSpace,
        offset: 451,
        len: 1,
    },
    Token {
        kind: Question2Eq,
        offset: 452,
        len: 3,
    },
    Token {
        kind: WhiteSpace,
        offset: 455,
        len: 1,
    },
    Token {
        kind: FatArrow,
        offset: 456,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 458,
        len: 5,
    },
    Token {
        kind: Question,
        offset: 463,
        len: 1,
    },
    Token {
        kind: Dot,
        offset: 464,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 465,
        len: 1,
    },
    Token {
        kind: Slash,
        offset: 466,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 467,
        len: 1,
    },
    Token {
        kind: SlashEq,
        offset: 468,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 470,
        len: 1,
    },
    Token {
        kind: RCurly,
        offset: 471,
        len: 1,
    },
    Token {
        kind: EOF,
        offset: 472,
        len: 1,
    },
]
